plugin/plugin.go

```diff
@@ -3,99 +3,44 @@ package plugin
 	"github.com/odpf/optimus/plugin/yaml"
 )
 
 func Initialize(pluginLogger hclog.Logger, arg ...string) error {
-	discoveredPlugins := DiscoverPlugins(pluginLogger)
-	pluginLogger.Debug(fmt.Sprintf("discovering binary plugins(%d)...", len(discoveredPlugins)))
-
-	// pluginMap is the map of plugins we can dispense.
-	pluginMap := map[string]plugin.Plugin{
-		models.PluginTypeBase:                     base.NewPluginClient(pluginLogger),
-		models.ModTypeDependencyResolver.String(): dependencyresolver.NewPluginClient(pluginLogger),
-	}
-
-	for _, pluginPath := range discoveredPlugins {
-		// we are core, start by launching the plugin processes
-		pluginClient := plugin.NewClient(&plugin.ClientConfig{
-			HandshakeConfig:  base.Handshake,
-			Plugins:          pluginMap,
-			Cmd:              exec.Command(pluginPath, arg...),
-			Managed:          true,
-			AllowedProtocols: []plugin.Protocol{plugin.ProtocolGRPC},
-			Logger:           pluginLogger,
-		})
-
-		// connect via GRPC
-		rpcClient, err := pluginClient.Client()
-		if err != nil {
-			return fmt.Errorf("client.Client(): %s: %w", pluginPath, err)
-		}
-
-		var baseClient models.BasePlugin
-		var drClient models.DependencyResolverMod
-
-		// request plugin as base
-		raw, err := rpcClient.Dispense(models.PluginTypeBase)
-		if err != nil {
-			pluginClient.Kill()
-			return fmt.Errorf("rpcClient.Dispense: %s: %w", pluginPath, err)
-		}
-		baseClient = raw.(models.BasePlugin)
-		baseInfo, err := baseClient.PluginInfo()
-		if err != nil {
-			return fmt.Errorf("failed to read plugin info: %s: %w", pluginPath, err)
-		}
-		pluginLogger.Debug("plugin connection established: ", baseInfo.Name)
-
-		if modSupported(baseInfo.PluginMods, models.ModTypeDependencyResolver) {
-			// create a client with dependency resolver mod
-			if rawMod, err := rpcClient.Dispense(models.ModTypeDependencyResolver.String()); err == nil {
-				drClient = rawMod.(models.DependencyResolverMod)
-				pluginLogger.Debug(fmt.Sprintf("%s mod found for: %s", models.ModTypeDependencyResolver, baseInfo.Name))
-
-				// cache name
-				drGRPCClient := rawMod.(*dependencyresolver.GRPCClient)
-				drGRPCClient.SetName(baseInfo.Name)
-			}
-		}
-
-		if err := models.PluginRegistry.Add(baseClient, drClient); err != nil {
-			return fmt.Errorf("PluginRegistry.Add: %s: %w", pluginPath, err)
-		}
-		pluginLogger.Debug("plugin ready: ", baseInfo.Name)
-	}
-
-	// first fetch binary plugins, then add yaml plugin for which binary is not added earlier
-	discoveredYamlPlugins := DiscoverPluginsGivenFilePattern(pluginLogger, yaml.Prefix, yaml.Suffix)
+	// fetch yaml plugins first, it holds detailed information about the plugin
+	discoveredYamlPlugins := discoverPluginsGivenFilePattern(pluginLogger, yaml.Prefix, yaml.Suffix)
 	pluginLogger.Debug(fmt.Sprintf("discovering yaml   plugins(%d)...", len(discoveredYamlPlugins)))
 	yaml.Init(models.PluginRegistry, discoveredYamlPlugins, pluginLogger)
```

Comment:
handle error. earlier yamlmod was optional.

Reply:
it's already coverd in binary Init